src/rust/inetstack/protocols/tcp/passive_open.rs

```diff
@@ -147,8 +166,32 @@ impl<N: NetworkRuntime> SharedPassiveSocket<N> {
         self.ready.pop(&yielder).await?
     }
 
+    // Closes the target socket.
+    pub fn close(&mut self) -> Result<(), Fail> {
+        self.state = State::Closed;
+        Ok(())
+    }
+
     async fn poll(mut self, yielder: Yielder) {
         loop {
+            match self.socket_rx.try_next() {
+                Ok(Some(socket)) => match self.connections.remove(&socket) {
```

Comment:
This particular check doesn't seem necessary to implement passive close. If the passive socket is closed, we will simply drop all packets headed to that address. So I would suggest that you do not need to introduce a mpsc channel here at all.

Reply:
Unfortunately, we need this, because the `recv_queue` is kept on the table and thus its drop logic does not run.